Readme.md

# koa-ratelimit

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![node version][node-image]][node-url]

[npm-image]: https://img.shields.io/npm/v/koa-ratelimit.svg?style=flat-square
[npm-url]: https://npmjs.org/package/koa-ratelimit
[travis-image]: https://img.shields.io/travis/koajs/ratelimit.svg?style=flat-square
[travis-url]: https://travis-ci.org/koajs/ratelimit
[node-image]: https://img.shields.io/badge/node.js-%3E=_0.11-red.svg?style=flat-square
[node-url]: http://nodejs.org/download/

 Rate limiter middleware for koa.

## Installation

```js
$ npm install koa-ratelimit
```

## Usage

The library exports 3 things:
- default: a factory accepting a `ratelimiter` adapter and which a function with options to build the middleware
- `redisRateLimit`: an helper function returning a middleware initialized with a redisAdapter
- `memoryRateLimit`: an helper function returning a middleware initialized with a memoryAdapter

## redisRateLimit

### Options

 - `db` redis connection instance
 - `max` max requests within `duration` [2500]
 - `duration` of limit in milliseconds [3600000]
 - `id` id to compare requests [ip]

### Example

```js
var ratelimit = require('koa-ratelimit').redisRateLimit;
var redis = require('redis');
var koa = require('koa');
var app = koa();

// apply rate limit

app.use(ratelimit({
  db: redis.createClient(),
  duration: 60000,
  max: 100,
  id: function (context) {
    return context.ip;
  }
}));

// response middleware

app.use(function *(){
  this.body = 'Stuff!';
});

app.listen(3000);
console.log('listening on port 3000');
```

## memoryRateLimit

### Options

 - `max` max requests within `duration` [2500]
 - `duration` of limit in milliseconds [3600000]
 - `id` id to compare requests [ip]

### Example

```js
var ratelimit = require('koa-ratelimit').memoryRateLimit;
var koa = require('koa');
var app = koa();

// apply rate limit

app.use(ratelimit({
  duration: 60000,
  max: 100,
  id: function (context) {
    return context.ip;
  }
}));

// response middleware

app.use(function *(){
  this.body = 'Stuff!';
});

app.listen(3000);
console.log('listening on port 3000');
```

## custom adapter

To learn more about what a custom adapter should be, please refer to the `ratelimiter` [documentation](https://github.com/marmelab/node-ratelimiter/tree/adapters)

### Options

 - `max` max requests within `duration` [2500]
 - `duration` of limit in milliseconds [3600000]
 - `id` id to compare requests [ip]

### Example

```js
var ratelimitFactory = require('koa-ratelimit');
var koa = require('koa');
var app = koa();
var myCustomAdapter = require('./myCustomAdapter');
// apply rate limit

app.use(ratelimitFactory(myCustomAdapter)({
  duration: 60000,
  max: 100,
  id: function (context) {
    return context.ip;
  }
}));

// response middleware

app.use(function *(){
 this.body = 'Stuff!';
});

app.listen(3000);
console.log('listening on port 3000');
```

## Responses

  Example 200 with header fields:

```
HTTP/1.1 200 OK
X-Powered-By: koa
X-RateLimit-Limit: 100
X-RateLimit-Remaining: 99
X-RateLimit-Reset: 1384377793
Content-Type: text/plain; charset=utf-8
Content-Length: 6
Date: Wed, 13 Nov 2013 21:22:13 GMT
Connection: keep-alive

Stuff!
```

  Example 429 response:

```
HTTP/1.1 429 Too Many Requests
X-Powered-By: koa
X-RateLimit-Limit: 100
X-RateLimit-Remaining: 0
X-RateLimit-Reset: 1384377716
Content-Type: text/plain; charset=utf-8
Content-Length: 39
Retry-After: 7
Date: Wed, 13 Nov 2013 21:21:48 GMT
Connection: keep-alive

Rate limit exceeded, retry in 8 seconds
```

## License

  MIT


History.md
3.0.0 / 2016-03-01
==================
* rewritten using es6 with babel
* uses new node/rate-limiter with adapters

2.1.0 / 2015-01-04
==================

  * Merge pull request #11 from seegno/support/flush-redis-before-testing
  * Merge pull request #9 from seegno/enhancement/add-id-verification
  * Flush redis limit keys before running tests
  * Add id verification to bypass ratelimiting

2.0.0 / 2015-01-02
==================

  * Merge pull request #10 from seegno/support/update-ratelimiter-2-0-0
  * Update `debug@2.1.1` and `ms@0.7.0`
  * Update recommended node version to >= 0.11.13
  * Update dev dependencies
  * Update ratelimiter@2.0.0
  * Replace `thunkify` by `thenify`.

1.1.0 / 2014-12-03
==================

  * Adds callback for defining own comparator ID. Fixes #3

1.0.3 / 2014-05-24
==================

 * fix: don't yield to downstream middleware if ratelimit is hit.

1.0.2 / 2014-05-22
==================

 * yield to next only when rate limit isn't hit

1.0.1 / 2013-11-29
==================

 * fix ms dep


src/index.spec.js
import koa from 'koa';
import request from 'supertest';
import 'should';
import redis from 'redis';

import { memoryRateLimit, redisRateLimit } from '.';

describe('ratelimit middleware with memoryAdapter', () => {
    const rateLimitDuration = 1000;
    const goodBody = 'Num times hit: ';

    describe('limit', () => {
        let guard;
        let app;

        const routeHitOnlyOnce = () => {
            guard.should.be.equal(1);
        };

        beforeEach(done => {
            app = koa();

            app.use(memoryRateLimit({
                duration: rateLimitDuration,
                max: 1,
            }));

            app.use(function* next() {
                guard++;
                this.body = goodBody + guard;
            });

            guard = 0;

            setTimeout(() => {
                request(app.listen())
                    .get('/')
                    .expect(200, goodBody + '1')
                    .expect(routeHitOnlyOnce)
                    .end(done);
            }, rateLimitDuration);
        });

        it('responds with 429 when rate limit is exceeded', done => {
            request(app.listen())
                .get('/')
                .expect('X-RateLimit-Remaining', 0)
                .expect(429)
                .end(done);
        });

        it('should not yield downstream if ratelimit is exceeded', done => {
            request(app.listen())
                .get('/')
                .expect(429)
                .end(() => {
                    routeHitOnlyOnce();
                    done();
                });
        });
    });

    describe('id', () => {
        it('should allow specifying a custom `id` function', done => {
            const app = koa();

            app.use(memoryRateLimit({
                max: 1,
                id: ctx => ctx.request.header.foo,
            }));

            request(app.listen())
                .get('/')
                .set('foo', 'bar')
                .expect(res => {
                    res.header['x-ratelimit-remaining'].should.equal('0');
                })
                .end(done);
        });

        it('should not limit if `id` returns `false`', done => {
            const app = koa();

            app.use(memoryRateLimit({
                id: () => false,
                max: 5,
            }));

            request(app.listen())
                .get('/')
                .expect(res => {
                    res.header.should.not.have.property('x-ratelimit-remaining');
                })
                .end(done);
        });

        it('should limit using the `id` value', done => {
            const app = koa();

            app.use(memoryRateLimit({
                max: 1,
                id: ctx => ctx.request.header.foo,
            }));

            app.use(function* next() {
                this.body = this.request.header.foo;
            });

            request(app.listen())
                .get('/')
                .set('foo', 'bar')
                .expect(200, 'bar')
                .end(() => {
                    request(app.listen())
                        .get('/')
                        .set('foo', 'biz')
                        .expect(200, 'biz')
                        .end(done);
                });
        });
    });
});

describe('ratelimit middleware with redis', () => {
    const db = redis.createClient();
    const rateLimitDuration = 1000;
    const goodBody = 'Num times hit: ';

    before(done => {
        db.keys('limit:*', (err, rows) => {
            rows.forEach(db.del, db);
        });

        done();
    });

    describe('limit', () => {
        let guard;
        let app;

        const routeHitOnlyOnce = () => {
            guard.should.be.equal(1);
        };

        beforeEach(done => {
            app = koa();

            app.use(redisRateLimit({
                db,
                duration: rateLimitDuration,
                max: 1,
            }));

            app.use(function* next() {
                guard++;
                this.body = goodBody + guard;
            });

            guard = 0;

            setTimeout(() => {
                request(app.listen())
                    .get('/')
                    .expect(200, goodBody + '1')
                    .expect(routeHitOnlyOnce)
                    .end(done);
            }, rateLimitDuration);
        });

        it('responds with 429 when rate limit is exceeded', done => {
            request(app.listen())
                .get('/')
                .expect('X-RateLimit-Remaining', 0)
                .expect(429)
                .end(done);
        });

        it('should not yield downstream if ratelimit is exceeded', done => {
            request(app.listen())
                .get('/')
                .expect(429)
                .end(() => {
                    routeHitOnlyOnce();
                    done();
                });
        });
    });

    describe('id', () => {
        it('should allow specifying a custom `id` function', done => {
            const app = koa();

            app.use(redisRateLimit({
                db,
                max: 1,
                id: ctx => ctx.request.header.foo,
            }));

            request(app.listen())
                .get('/')
                .set('foo', 'bar')
                .expect(res => {
                    res.header['x-ratelimit-remaining'].should.equal('0');
                })
                .end(done);
        });

        it('should not limit if `id` returns `false`', done => {
            const app = koa();

            app.use(redisRateLimit({
                db,
                id: () => false,
                max: 5,
            }));

            request(app.listen())
                .get('/')
                .expect(res => {
                    res.header.should.not.have.property('x-ratelimit-remaining');
                })
                .end(done);
        });

        it('should limit using the `id` value', done => {
            const app = koa();

            app.use(redisRateLimit({
                db,
                max: 1,
                id: ctx => ctx.request.header.foo,
            }));

            app.use(function* next() {
                this.body = this.request.header.foo;
            });

            request(app.listen())
                .get('/')
                .set('foo', 'bar')
                .expect(200, 'bar')
                .end(() => {
                    request(app.listen())
                        .get('/')
                        .set('foo', 'biz')
                        .expect(200, 'biz')
                        .end(done);
                });
        });
    });
});


src/index.js
/**
 * Module dependencies.
 */

import debugFactory from 'debug';
import Limiter, { memoryAdapter, redisAdapter } from 'ratelimiter';
import ms from 'ms';

const debug = debugFactory('koa-ratelimit');

/**
 * Initialize ratelimit middleware with the given `opts`:
 *
 * - `duration` limit duration in milliseconds [1 hour]
 * - `max` max requests per `id` [2500]
 * - `db` database connection
 * - `id` id to compare requests [ip]
 *
 * @param {Object} opts
 * @param {Object} adapter [memoryAdapter]
 * @return {Function}
 * @api public
 */
const middlewareFactory = (adapter = memoryAdapter()) => opts => function* middleware(next) {
    const id = opts.id ? opts.id(this) : this.ip;

    if (id === false) return yield * next;

    // initialize limiter
    const limiter = new Limiter({ ...opts, id }, adapter);

    // check limit
    const limit = yield limiter.newHit();

    // check if current call is legit
    const remaining = limit.remaining > 0 ? limit.remaining - 1 : 0;

    // header fields
    this.set('X-RateLimit-Limit', limit.total);
    this.set('X-RateLimit-Remaining', remaining);
    this.set('X-RateLimit-Reset', limit.reset);

    debug('remaining %s/%s %s', remaining, limit.total, id);
    if (limit.remaining) return yield * next;

    const delta = (limit.reset * 1000) - Date.now() | 0;
    const after = limit.reset - (Date.now() / 1000) | 0;
    this.set('Retry-After', after);
    this.status = 429;
    this.body = 'Rate limit exceeded, retry in ' + ms(delta, {
        long: true,
    });
};

export default middlewareFactory;

export const memoryRateLimit = opts => middlewareFactory(memoryAdapter())(opts);

export const redisRateLimit = opts => {
    if (!opts || !opts.db) {
        throw new Error('opts.db is required');
    }

    return middlewareFactory(redisAdapter(opts.db))(opts);
};
